crates/storage/provider/src/change.rs

```diff
@@ -0,0 +1,1325 @@
+    pub fn detach_lower_part_at(&mut self, block_number: BlockNumber) -> Option<Self> {
+        let last_block = self.last_block();
+        let first_block = self.first_block;
+        if block_number >= last_block {
+            return None
+        }
+        if block_number < first_block {
+            return Some(Self::default())
+        }
+
+        // detached number should be included so we are adding +1 to it.
+        // for example if block number is same as first_block then
+        // number of detached block shoud be 1.
+        let num_of_detached_block = (block_number - first_block) + 1;
+
+        let mut detached_bundle_state: BundleState = self.clone();
+        detached_bundle_state.revert_to(block_number);
+
+        // split is done as [0, num) and [num, len]
+        let (_, this) = self.receipts.split_at(num_of_detached_block as usize);
+
+        self.receipts = this.to_vec().clone();
+        self.bundle
+            .detach_lower_part_reverts(num_of_detached_block as usize)
+            .expect("there should be detachments");
+
+        self.first_block = block_number + 1;
+
+        Some(detached_bundle_state)
+    }
+
+    /// Extend one state from another
+    ///
+    /// For state this is very sensitive opperation and should be used only when
+    /// we know that other state was build on top of this one.
+    /// In most cases this would be true.
+    pub fn extend(&mut self, other: Self) {
+        self.bundle.extend(other.bundle);
+        self.receipts.extend(other.receipts);
+    }
+
+    /// Write bundle state to database.
+    ///
+    /// `omit_changed_check` should be set to true of bundle has some of it data
+    /// detached, This would make some original values not known.
+    pub fn write_to_db<'a, TX: DbTxMut<'a> + DbTx<'a>>(
+        mut self,
+        tx: &TX,
+        omit_changed_check: bool,
+    ) -> Result<(), DatabaseError> {
+        // write receipts
+        let mut receipts_cursor = tx.cursor_write::<tables::Receipts>()?;
+        let mut next_number = receipts_cursor.last()?.map(|(i, _)| i + 1).unwrap_or_default();
+        for block_receipts in self.receipts.into_iter() {
+            for receipt in block_receipts {
+                receipts_cursor.append(next_number, receipt)?;
+                next_number += 1;
+            }
+        }
+        let reverts = self.bundle.take_reverts();
+        StateReverts(reverts).write_to_db(tx, self.first_block)?;
+
+        StateChange(self.bundle.into_plain_state_sorted(omit_changed_check)).write_to_db(tx)?;
+
+        Ok(())
+    }
+}
+
+/// Revert of the state.
+#[derive(Default)]
+pub struct StateReverts(pub RevmReverts);
+
+impl From<RevmReverts> for StateReverts {
+    fn from(revm: RevmReverts) -> Self {
+        Self(revm)
+    }
+}
+
+impl StateReverts {
+    /// Write reverts to database.
+    ///
+    /// Note:: Reverts will delete all wiped storage from plain state.
+    pub fn write_to_db<'a, TX: DbTxMut<'a> + DbTx<'a>>(
```

Comment:
the naming is very confusing to me, because writing reverts results in wiped storage, because this does not actually write anything and instead deletes?

Reply:
This function both writes/cleans up empty plainstate, but it also appends (the reverts) to the changesets tables. Makes me wonder if we should call reverts Prestate's or something.

Reply:
Created the issue for this as it is debt from before revm state: https://github.com/paradigmxyz/reth/issues/4620